### DISCUSSION TITLE
tutorials/shaders/your_first_shader/your_second_3d_shader

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#95] [User-contributed notes]
# tutorials/shaders/your_first_shader/your_second_3d_shader

From a high-level, what Godot does is give the user a bunch of parameters that can be optionally set ( AO, SSS_Strength, RIM, etc.). These parameters correspond to different complex effects (Ambien...

https://docs.godotengine.org/en/4.3/tutorials/shaders/your_first_shader/your_second_3d_shader.html

**Comment by ChrisDWhitehead** [comment]
I have been attempting to follow along with this tutorial but  the results I get are nothing like what is pictured. It would be very helpful if there was a complete code listing at the end, like in the 2D Shader tutorial. I suspect there may be code missing in the running explaination.

**Reply by Catley94** [reply]
I have posted my code below which works, I have had to adapt it slightly and I've made a modification with the 'calmness' of the water, though you are welcome to ignore this and replace the variable with the magic number of 10.0.

**Reply by mrlem** [reply]
Things that I've found missing between the 2 tutorials:
* you should resize your plane from 2x2 (first tutorial) to 10x10 (assumed in the second tutorial)
* you should make your noise texture seamless to avoid having a cyclical "crest" at the edge of the texture

**Reply by Catley94** [reply]
Yes! Though didn't know about the second, will have a look at that later. Thanks!

**Comment by Catley94** [comment]
I also am struggling with this - there are two things that I have noticed thus far, when I went from part 1 => part 2, my peaks were sharper than the first picture in part 2, though I ignored this. 
But where I really got stuck was at TIME, the first mention of vertex function you showed was:
```glsl
void vertex() {
  vec2 pos = VERTEX.xz;
  float k = height(pos, TIME);
  VERTEX.y = k;
}
```

However before I made any changes, my looks like:
```glsl
void vertex() {
	tex_position = VERTEX.xz / 2.0 + 0.5;
	float height = texture(noise, tex_position).x;
	VERTEX.y += height * height_scale;
}
```

I'm figuring this out now - but I feel there's a disconnection between what was, and what is. (Though this could also be me just in too deep currently).
Will continue on and post my code at the end to see if others got the same :-)

**Reply by Calinou** [reply]
PS: Code blocks should use triple backticks like this (with an optional language name for syntax highlighting):

\`\`\`glsl
code here
\`\`\`

I edited your post accordingly, but remember to do this in the future :slightly_smiling_face:

**Reply by Catley94** [reply]
Thank you! 😊

**Comment by Catley94** [comment]
This is my final code, though I have modified it slightly to 'tame' the waters, as my implementation following the tutorial was very choppy and fast. 

The calmness variable goes from 1 - 100, though I would stick between 20 and 60 (this number depends on the amount of waves you create with `d += wave((position - time) * 0.6) * 0.2; //Or similar`).

The below code has nicely resulted in a section of water that is relatively calm with multiple layers of water moving in different directions.

```glsl
shader_type spatial;
render_mode specular_toon;

uniform float height_scale = 0.5;
uniform sampler2D noise;
uniform sampler2D normalmap;
uniform float calmness = 50.0; // 1 - 100
uniform int num_of_additional_waves_in_x_direction = 2; // May be labeled as the wrong direction
uniform int num_of_additional_waves_in_z_direction = 2; // May be labeled as the wrong direction

varying vec2 tex_position;

float wave(vec2 position) {
	position += texture(noise, position / calmness).x * 2.0 - 1.0;
	vec2 wv = 1.0 - abs(sin(position));
	return pow(1.0 - pow(wv.x * wv.y, 0.65), 4.0);
}

float height(vec2 position, float time) {
	// Change your MeshInstance3D PlaneMesh to 10 x 10

	float d = wave((position + time) * 0.4) * 0.3;
	
	// Give the user control over how many waves and in what direction
	
	//for(int i = 0; i < num_of_additional_waves_in_x_direction; i++) {
		////Randomise magic number 0.3 and 0.3 to a range beteen 0.1 and 1
		//d += wave((position + time) * 0.3) * 0.3;
	//}
	//
	//for(int i = 0; i < num_of_additional_waves_in_z_direction; i++) {
		////Randomise magic number 0.3 and 0.3 to a range beteen 0.1 and 1
		//d += wave((position - time) * 0.3) * 0.3;
	//}
	
	// Statically create the waves
	
	d += wave((position - time) * 0.3) * 0.3;
	d += wave((position + time) * 0.5) * 0.2;
	d += wave((position - time) * 0.6) * 0.2;
	return d;
}


void vertex() {
	// Called for every vertex the material is visible on.	

	//Do I change the below to height(tex_position, TIME); ?
	// Version - 1
	//tex_position = VERTEX.xz / 5.0 + 0.5;
	//float _height = texture(noise, tex_position).x;
	
	//VERTEX.y += _height * height_scale;
	
	
	vec2 pos = VERTEX.xz;
	float k = height(pos, TIME);

	// Version - 2
  	VERTEX.y = k;
	
	NORMAL = normalize(vec3(k - height(pos + vec2(0.1, 0.0), TIME), 0.1, k - height(pos + vec2(0.0, 0.1), TIME)));
	
	
}

void fragment() {
	// Called for every pixel the material is visible on.
	
	float fresnel = sqrt(1.0 - dot(NORMAL, VIEW));
	NORMAL_MAP = texture(normalmap, tex_position).xyz;
	
	
	RIM = 0.2;
	METALLIC = 0.0;
	ROUGHNESS = 0.01 * (1.0 - fresnel);
	ALBEDO = vec3(0.01, 0.03, 0.05) +  + (0.1 * fresnel);
	
}

//void light() {
	// Called for every pixel for every light affecting the material.
	// Uncomment to replace the default light processing function with this one.
//}```

**Reply by mrlem** [reply]
I'd say in your code you should remove the `NORMAL_MAP` in `fragment()`, as the normal is now calculated in `vertex()`

**Reply by Catley94** [reply]
Thanks for this, I'm still learning shaders and thought they were different things haha. Currently working my way through TheBookofShaders hoping this will help :)

**Comment by Vosburgh** [comment]
This tutorial needs a do over, shaders are already difficult enough. Why should the tutorial skip steps and leave guess work for what the author did?

Atleast leave us something we can cross reference against.

**Reply by Calinou** [reply]
This is being tracked in https://github.com/godotengine/godot-docs/issues/8077.

**Comment by and-then-there-was-ben** [comment]
I put together a clean, heavily-commented, and hopefully more intuitive shader based on the comments above. I included all necessary settings for the mesh, light, and noise map in order to recreate my own preview scene. Even the light's position is given to avoid different results. The base code is close enough to the original to follow.

+ I tried to expand a lot of the math into named parameters so you can visually understand what they're doing, then jump back in and reverse engineer the code, rather than trying to follow it from scratch.
+ I broke out the displacement effects into three properties. WaveHeight controls the overall intensity, allowing you to return to a perfectly flat plane. The other two control the vertex offset (displacement) and the normal map's effect. The latter is purely to see how much each contributes to the overall effect.
+ The waves can be individually toggled as Catley had experimented with.
+ There are three lighting parameters as well. Ambient lighting, passed through the unused AO channel, applies a blanket effect over the entire surface, effectively controlling all ambient lighting to the shader. Set it to 0 and the surface becomes black. You can bind that to fresnel so the two scale nicely together for really simple night-time water effects.

This shader isn't meant to demonstrate what is "correct". It's meant as a toy that explains itself.

```glsl
shader_type spatial;
render_mode specular_toon;

// ########## SETUP ########### 
// Keep only the first noise map from the previous shader.
// Remove all other properties and clear the vertex() and fragment() shaders.

// Noise Map = NoiseTexture2d
// Noise Type = Simplex Smooth
// Seamless = true
// As Normal Map = false

// Primitive = Plane
// Size = (10m x 10m)
// Subdivisions = (640 x 640)
	
// Lighting = OmniLight3D
// Range = 25.0
// Energy = 2.0
// Transform = (0, 3, -5) // Adjust height for some quick variations.
// Disable preview sunlight.

// Base wave settings
uniform sampler2D noise;
uniform float waveHeight = 1.0; // Zero makes a flat plane.
uniform float displacementScale = 0.75; // Only affects vertex displacement.
uniform float normalMapIntensity = 1.0; // As a contrast to the vertex displacement. Should be 1.0
uniform float waveTile = 0.5;
uniform float waveSpeed = 0.5;

// Toggle waves
uniform bool waveA = true;
uniform bool waveB = true;
uniform bool waveC = true;
uniform bool waveD = true;

// Stylize the waves
uniform float contrast = 1.0; // Sort of a contrast, sort of a complexity/octave.
uniform float ridgePucker = 1.0; // Demonstrates how the Pow() method behaves.

// Explore lighting layers
uniform float ambientIntensity = 1.0; // Basically the overall light as implemented. Scales fresnel, but not color.
uniform float fresnelIntensity = 0.1;
uniform float colorIntensity = 1.0;

float wave(vec2 position) {
	position += texture(noise, position / (10.0/waveTile)).x * (2.0 * contrast); // No clear reason for the (-1) offset.
	vec2 wv = 1.0 - abs(sin(position));
	return pow(1.0 - pow(wv.x * wv.y, 0.65 * ridgePucker), 4.0) * waveHeight;
}

// Add several wave patterns on top of each other to form a single offset mask.
// Toggle them using the exposed bool parameters.
float height(vec2 position, float time) {
	time = time * waveSpeed;
	float h = 0.0;
	if(waveA){ h = wave((position + time) * 0.4) * 0.3; }
	if(waveB){ h += wave((position - time) * 0.3) * 0.3; }
	if(waveC){ h += wave((position + time) * 0.5) * 0.2; }
	if(waveD){ h += wave((position - time) * 0.6) * 0.2; }
	return h;
}

void vertex() {
	vec2 pos = VERTEX.xz;
	float k = height(pos, TIME);
	VERTEX.y = k * displacementScale;
	vec3 waveNormals = normalize(vec3(k - height(pos + vec2(0.1, 0.0), TIME), 0.1, k - height(pos + vec2(0.0, 0.1), TIME)));
	// This lerps from NORMAL to waveNormals based on the intensity value.
	NORMAL = (NORMAL * (1.0 - normalMapIntensity)) + (waveNormals * normalMapIntensity);
}

void fragment() {
	// A fresnel angle is a mask describing how perpendicular a surface angle is to your viewing angle.
	// It can be used to create a rim light, as we do here, or cheap SSS effects.
	float fresnel = sqrt(1.0 - dot(NORMAL, VIEW)) * fresnelIntensity;// * ambientIntensity;
	RIM = 0.2;
	METALLIC = 0.0;
	ROUGHNESS = 0.01 * (1.0 - fresnel);
	ALBEDO = (vec3(0.01, 0.03, 0.05) * colorIntensity) + fresnel;
	AO = ambientIntensity; // Set to zero to block all ambient lighting.
}

// Parameters are set with safe defaults. Try changing a few at a time, then reset them.

// Try WaveHeight(0.2), Displacement(0.1), Contrast(5) for a windy river-like effect.
// Try Ambient(0.05), Fresnel(0.02) and lower the omni light to Y(1.75) for a night sky with just the moon's reflection.
// Try Ambient(1), Fresnel(-0.1), Color(0.2) for an oily black water (artificial).
```
